buildSrc/src/main/java/org/elasticsearch/gradle/VersionCollection.java

```diff
@@ -0,0 +1,331 @@
+        Set<Version> notReallyReleased = new HashSet<>(getReleased());
+        notReallyReleased.removeAll(authoritativeReleasedVersions);
+        if (notReallyReleased.isEmpty() == false) {
+            throw new IllegalStateException(
+                "out-of-date released versions" +
+                    "\nFollowing versions are not really released, but the build thinks they are: " + notReallyReleased
+            );
+        }
+
+        Set<Version> incorrectlyConsideredUnreleased = new HashSet<>(authoritativeReleasedVersions);
+        incorrectlyConsideredUnreleased.retainAll(getUnreleased());
+        if (incorrectlyConsideredUnreleased.isEmpty() == false) {
+            throw new IllegalStateException(
+                "out-of-date released versions" +
+                    "\nBuild considers versions unreleased, " +
+                    "but they are released according to an authoritative source: " + incorrectlyConsideredUnreleased +
+                    "\nThe next versions probably needs to be added to Version.java (CURRENT doesn't count)."
+            );
+        }
+    }
+
+    private List<Version> getReleased() {
+        List<Version> unreleased = getUnreleased();
+        return groupByMajor.values().stream()
+            .flatMap(Collection::stream)
+            .filter(each -> unreleased.contains(each) == false)
+            .sorted()
+            .collect(Collectors.toList());
+    }
+
+    public List<Version> getIndexCompatible() {
+        return unmodifiableList(
+            groupByMajor.values().stream()
+                .flatMap(Collection::stream)
+                .filter(version -> version.equals(currentVersion) == false)
+                .sorted()
+                .collect(Collectors.toList())
+        );
+    }
+
+    public List<Version> getWireCompatible() {
+        Version lastPreviousMinor = getLatestVersionByKey(groupByMajor, currentVersion.getMajor() - 1);
+        Version firstPreviousMinor = new Version(lastPreviousMinor.getMajor(), lastPreviousMinor.getMinor(), 0);
+        return unmodifiableList(
+            groupByMajor.values()
+                .stream()
+                .flatMap(Collection::stream)
+                .filter(version -> version.compareTo(firstPreviousMinor) >= 0)
+                .filter(version -> version.equals(currentVersion) == false)
+                .sorted()
+                .collect(Collectors.toList())
+        );
+    }
+
+    public List<Version> getUnreleasedIndexCompatible() {
+        List<Version> unreleasedIndexCompatible = new ArrayList<>(getIndexCompatible());
+        unreleasedIndexCompatible.retainAll(getUnreleased());
+        return unmodifiableList(unreleasedIndexCompatible);
+    }
+
+    public List<Version> getUnreleasedWireCompatible() {
+        List<Version> unreleasedWireCompatible = new ArrayList<>(getWireCompatible());
+        unreleasedWireCompatible.retainAll(getUnreleased());
```

Comment:
retainAll with 2 lists is going to be quadratic. The unreleased list should be sorted. The same logic suggested above on finding index/wire compatible can be used here.

Reply:
I really like how well this reads, and the number of versions is finite in practice.